### ISSUE TITLE
Request support for eMylo Wifi Smart Power Meter

### ISSUE THREAD

**Issue by fjvva** [make-all/tuya-local#529]
Hello,

It would be great if you could add support for this smart power meter. Thank you in advance, and please let me know if i can help in any way!

# Log Message

This is the log from "Get device details" on tuya IOT cloud:

```
{
  "result": {
    "active_time": 1668161297,
    "biz_type": 0,
    "category": "zndb",
    "create_time": 1668161297,
    "icon": "smart/icon/ay1536897190133VVeLx/32346148448dd9829abb97ce317c5e82.png",
    "id": "X",
    "ip": "X",
    "lat": "X",
    "local_key": "X",
    "lon": "X",
    "model": "",
    "name": "Wifi Smart Meter",
    "online": true,
    "owner_id": "X",
    "product_id": "nqbs1onwskmmmaac",
    "product_name": "wifi smart meter",
    "status": [
      {
        "code": "forward_energy_total",
        "value": 949095
      },
      {
        "code": "phase_a",
        "value": "{\"electricCurrent\":14.6,\"power\":3.322,\"voltage\":228.2}"
      },
      {
        "code": "fault",
        "value": 0
      },
      {
        "code": "switch_prepayment",
        "value": false
      },
      {
        "code": "balance_energy",
        "value": 0
      },
      {
        "code": "clear_energy",
        "value": false
      },
      {
        "code": "charge_energy",
        "value": 0
      },
      {
        "code": "switch",
        "value": true
      },
      {
        "code": "alarm_set_2",
        "value": "[{\"alarmCode\":\"overcurrent\",\"doAction\":false,\"threshold\":\"30\"},{\"alarmCode\":\"ammeter_overvoltage\",\"doAction\":false,\"threshold\":\"245\"},{\"alarmCode\":\"under_voltage\",\"doAction\":false,\"threshold\":\"200\"},{\"alarmCode\":\"insufficient_balance\",\"doAction\":false,\"threshold\":\"30\"}]"
      },
      {
        "code": "event_clear",
        "value": false
      }
    ],
    "sub": false,
    "time_zone": "+01:00",
    "uid": "X",
    "update_time": 1668182241,
    "uuid": "X"
  },
  "success": true,
  "t": 1678565168725,
  "tid": "X"
}
```

The log from HA shows only this:

```
   Device matches None with quality of 0%. DPS: {"updated_at": 1678563322.8459601, "1": 948934, "10": 0, "12": false, "13": 0, "16": true}
    Device matches None with quality of 0%. DPS: {"updated_at": 1678563373.6594088, "1": 948938, "10": 0, "12": false, "13": 0, "16": true}
    Report this to https://github.com/make-all/tuya-local/issues/
    Device matches ir_remote_sensors with quality of 67%. DPS: {"updated_at": 1678563578.6888258, "1": true, "101": "2", "102": 0}
    Device matches None with quality of 0%. DPS: {"updated_at": 1678565714.7939818, "1": 949145, "10": 0, "12": false, "13": 0, "16": true}
```

# Information about DPS mappings
```

{
  "result": {
    "category": "zndb",
    "functions": [
      {
        "code": "switch_prepayment",
        "dp_id": 12,
        "type": "Boolean",
        "values": "{}"
      },
      {
        "code": "clear_energy",
        "dp_id": 14,
        "type": "Boolean",
        "values": "{}"
      },
      {
        "code": "charge_energy",
        "dp_id": 15,
        "type": "Integer",
        "values": "{\"unit\":\"kW·h\",\"min\":0,\"max\":50000,\"scale\":0,\"step\":1}"
      },
      {
        "code": "switch",
        "dp_id": 16,
        "type": "Boolean",
        "values": "{}"
      },
      {
        "code": "alarm_set_2",
        "dp_id": 17,
        "type": "Json",
        "values": "{}"
      },
      {
        "code": "event_clear",
        "dp_id": 20,
        "type": "Boolean",
        "values": "{}"
      }
    ],
    "status": [
      {
        "code": "forward_energy_total",
        "dp_id": 1,
        "type": "Integer",
        "values": "{\"unit\":\"kW·h\",\"min\":0,\"max\":99999999,\"scale\":2,\"step\":1}"
      },
      {
        "code": "phase_a",
        "dp_id": 6,
        "type": "Json",
        "values": "{}"
      },
      {
        "code": "fault",
        "dp_id": 10,
        "type": "Bitmap",
        "values": "{\"label\":[\"ov_cr\",\"unbalance_alarm\",\"ov_vol\",\"undervoltage_alarm\",\"miss_phase_alarm\",\"outage_alarm\",\"magnetism_alarm\",\"terminal_alarm\",\"cover_alarm\",\"credit_alarm\",\"no_balance_alarm\",\"battery_alarm\",\"meter_hardware_alarm\"]}"
      },
      {
        "code": "switch_prepayment",
        "dp_id": 12,
        "type": "Boolean",
        "values": "{}"
      },
      {
        "code": "balance_energy",
        "dp_id": 13,
        "type": "Integer",
        "values": "{\"unit\":\"kW·h\",\"min\":0,\"max\":99999999,\"scale\":2,\"step\":1}"
      },
      {
        "code": "clear_energy",
        "dp_id": 14,
        "type": "Boolean",
        "values": "{}"
      },
      {
        "code": "charge_energy",
        "dp_id": 15,
        "type": "Integer",
        "values": "{\"unit\":\"kW·h\",\"min\":0,\"max\":50000,\"scale\":0,\"step\":1}"
      },
      {
        "code": "switch",
        "dp_id": 16,
        "type": "Boolean",
        "values": "{}"
      },
      {
        "code": "alarm_set_2",
        "dp_id": 17,
        "type": "Json",
        "values": "{}"
      },
      {
        "code": "event_clear",
        "dp_id": 20,
        "type": "Boolean",
        "values": "{}"
      }
    ]
  }
}
```

Other DPs:
101 - Internal Expansion command (whatever that is)

# Product ID

```
{
  "result": {
    "active_time": 1668161297,
    "category": "zndb",
    "category_name": "Smart Electric Meter",
    "create_time": 1668161297,
    "gateway_id": "",
    "icon": "smart/icon/ay1536897190133VVeLx/32346148448dd9829abb97ce317c5e82.png",
    "id": "X",
    "ip": "X",
    "lat": "X",
    "local_key": "X",
    "lon": "X",
    "model": "",
    "name": "Wifi Smart Meter",
    "online": true,
    "owner_id": "X",
    "product_id": "nqbs1onwskmmmaac",
    "product_name": "wifi smart meter",
    "sub": false,
    "time_zone": "+01:00",
    "update_time": 1668182241,
    "uuid": "X"
  },
  "success": true,
  "t": 1678566312377,
  "tid": "X"
}
```

# Information about how the device functions

Here is the homepage of the product:
[http://emyloworld.com/products/details-58-223.html](http://emyloworld.com/products/details-58-223.html)

Here is an amazon page with more details and real captures of the information displayed on tuya:
[https://www.amazon.de/-/en/1-Phase-Digital-Wireless-Control-Smartphone/dp/B083BFGHPQ](https://www.amazon.de/-/en/1-Phase-Digital-Wireless-Control-Smartphone/dp/B083BFGHPQ)

Basically, it monitors power, has a few features to set limiters on (no leak on this specific model though) and has an internal relay that can be used to manually disconnect the power it forwards, or that will automatically open the circuit if the limiters set are reached or exceeded.

**Comment by mutozorro** [comment]
Hi make-all, thank you for your support.
The integration is partially working: it is able to retrieve the Total Energy but not the measure of Power, current, and voltage

...
    "product_name": "wifi smart meter",
    "status": [
      {
        "code": "forward_energy_total",
        "value": 45284
      },
      {
        "code": "phase_a",
        "value": "{\"electricCurrent\":7.26,\"power\":1.498,\"voltage\":232.0}"
      },
....

Please, let me know if I can help to improve this integration.

**Comment by mutozorro** [comment]
![image](https://user-images.githubusercontent.com/121493794/226063880-a893c6c2-728e-4e67-b0e7-581cc733ffcd.png)

**Comment by make-all** [comment]
The device does not expose current, voltage and power as separate attributes. They are encoded in the phase_a attribute by some undocumented encoding method. phase_a is exposed as an extra attribute on the energy sensor if you want to have a look at it.

**Comment by mutozorro** [comment]
Hi Jason,
attached are a couple of device configuration files for Emylo and TOMZN wifi intelligent energy meters.They work on my devices with the limitations you highlighted in the previous email.
I've just modified the two configuration files already present in the distribution.
Hope can help a little bit.


    Il sabato 18 marzo 2023 15:01:28 CET, Jason Rumney ***@***.***> ha scritto:  
 
 


Closed #529 as completed.
